### PR TITLE
[FuncSpec] Query SCCPSolver in more places

### PR DESCRIPTION
**Pull request by hazzlim**
When traversing the use-def chain of an Argument in a candidate specialization, also query the SCCPSolver to see if a Value is constant. This allows us to better estimate the codesize savings of a candidate in the presence of instructions that are a user of the argument we are estimating savings for which also use arguments that have been found constant by IPSCCP.

Similarly when estimating the dead basic blocks from branch and switch instructions which become constant, also query the SCCPSolver to see if a predecessor is unreachable.

### CODE REVIEW

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM generally.

File: llvm/lib/Transforms/IPO/FunctionSpecialization.cpp [hunk 1]
```diff
@@ -312,10 +317,10 @@ bool InstCostVisitor::discoverTransitivelyIncomingValues(
 
       // Disregard self-references and dead incoming values.
       if (auto *Inst = dyn_cast<Instruction>(V))
-        if (Inst == PN || DeadBlocks.contains(PN->getIncomingBlock(I)))
```
[the remark this answers is not in the document]
Comment (david-arm) [reply llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:315]: Would that suggest a bug then? Is there a way for the solver InstCostVisitor to reuse the same code that finds a basic block dead?

[the remark this answers is not in the document]
Comment (hazzlim) [reply llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:315]: It's maybe confusing that this function is named the same as the `SCCPSolver` member function, but this change is to make us dispatch to `InstCostVisitor::isBlockExecutable`, which checks both the `SCCPSolver` and DeadBlocks:
https://github.com/llvm/llvm-project/blob/2de3d00edc614cdce4b30b68be564a7a3123dbf9/llvm/include/llvm/Transforms/IPO/FunctionSpecialization.h#L179-L181

[the remark this answers is not in the document]
Comment (labrinea) [reply llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:315]: > Would that suggest a bug then?

No it's not a bug.

>  Is there a way for the solver and InstCostVisitor to reuse the same code that finds a basic block dead?

I am not sure I understand the question. The InstCostVisitor is a simplified version of the Solver for the pursposes of estimating the profitability of specialization. If the specialization is performed the Solver will end up matching the findings of InstCostVisitor.

[the remark this answers is not in the document]
Comment (labrinea) [reply llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:315]: > It's maybe confusing that this function is named the same as the SCCPSolver member function, but this change is to make us dispatch to InstCostVisitor::isBlockExecutable

Ah right, excuse me I had forgotten that. Then it makes sense. Can you use that function inside canEliminateSuccessor then?

[the remark this answers is not in the document]
Comment (hazzlim) [reply llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:315]: > Ah right, excuse me I had forgotten that. Then it makes sense. Can you use that function inside canEliminateSuccessor then?

Yep, will do this as well as updating things to be non-static/member functions, as per other comments.

[the remark this answers is not in the document]
Comment (hazzlim) [reply llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:315]: > Can you use that function inside canEliminateSuccessor then?

File: llvm/lib/Transforms/IPO/FunctionSpecialization.cpp [hunk 2]
```diff
@@ -91,12 +91,14 @@ static cl::opt<bool> SpecializeLiteralConstant(
         "argument"));
 
 bool InstCostVisitor::canEliminateSuccessor(BasicBlock *BB, BasicBlock *Succ,
-                                         DenseSet<BasicBlock *> &DeadBlocks) {
+                                            DenseSet<BasicBlock *> &DeadBlocks,
+                                            const SCCPSolver &Solver) {
```
[the remark this answers is not in the document]
Comment (hazzlim) [reply llvm/lib/Transforms/IPO/FunctionSpecialization.cpp]: Good point - updated to be non-static (and const).

File: llvm/lib/Transforms/IPO/FunctionSpecialization.cpp [hunk 3]
```diff
@@ -135,15 +137,18 @@ Cost InstCostVisitor::estimateBasicBlocks(
...
       if (isBlockExecutable(SuccBB) &&
-          canEliminateSuccessor(BB, SuccBB, DeadBlocks))
+          canEliminateSuccessor(BB, SuccBB, DeadBlocks, Solver))
         WorkList.push_back(SuccBB);
   }
   return CodeSize;
 }
 
-static Constant *findConstantFor(Value *V, ConstMap &KnownConstants) {
+static Constant *findConstantFor(Value *V, ConstMap &KnownConstants,
```
[the remark this answers is not in the document]
Comment (hazzlim) [reply llvm/lib/Transforms/IPO/FunctionSpecialization.cpp]: Good point - done.

File: llvm/lib/Transforms/IPO/FunctionSpecialization.cpp [hunk 4]
```diff
@@ -347,10 +352,10 @@ Constant *InstCostVisitor::visitPHINode(PHINode &I) {
 
     // Disregard self-references and dead incoming values.
     if (auto *Inst = dyn_cast<Instruction>(V))
-      if (Inst == &I || DeadBlocks.contains(I.getIncomingBlock(Idx)))
```
[the remark this answers is not in the document]
Comment (hazzlim) [reply llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:350]: I've left this as is, as per discussion above. I have added an assert just before inserting into DeadBlocks that the Solver thinks the block is executable.

**Review by labrinea — APPROVED** [review]
Looks good, thanks!

**Comment by hazzlim** [comment]
Thank you for reviewing! :)

**Comment by FlameTop** [comment]
We are currently experiencing assertion failures after this commit landed. A reproducer is 

https://godbolt.org/z/oTjdex7Gh

which is failing with the head of trunk at this date.

**Comment by davemgreen** [comment]
It looks like it doesn't like the metadata on the strict-fp call args. We should be able to skip that when looking at the operands.

**Comment by davemgreen** [comment]
#124284 hopefully addresses this. Thanks.

**Pull request merged**